cluster/src/main/java/io/atomix/cluster/messaging/impl/AbstractClientConnection.java

```diff
@@ -0,0 +1,183 @@
+      }
+    } else {
+      log.debug("Received a reply for message id:[{}] but was unable to locate the request handle", message.id());
+    }
+  }
+
+  /**
+   * Adds a reply time to the history.
+   *
+   * @param type      the message type
+   * @param replyTime the reply time to add to the history
+   */
+  private void addReplyTime(String type, long replyTime) {
+    DescriptiveStatistics samples = replySamples.get(type);
+    if (samples == null) {
+      samples = replySamples.computeIfAbsent(type, t -> new SynchronizedDescriptiveStatistics(WINDOW_SIZE));
+    }
+    samples.addValue(replyTime);
+  }
+
+  /**
+   * Returns the timeout in milliseconds for the given timeout duration
+   *
+   * @param type    the message type
+   * @param timeout the timeout duration or {@code null} if the timeout is dynamic
+   * @return the timeout in milliseconds
+   */
+  private long getTimeoutMillis(String type, Duration timeout) {
+    return timeout != null ? timeout.toMillis() : computeTimeoutMillis(type);
+  }
+
+  /**
+   * Computes the timeout for the next request.
+   *
+   * @param type the message type
+   * @return the computed timeout for the next request
+   */
+  private long computeTimeoutMillis(String type) {
+    DescriptiveStatistics samples = replySamples.get(type);
+    if (samples == null || samples.getN() < MIN_SAMPLES) {
+      return MAX_TIMEOUT_MILLIS;
+    }
+    return Math.min(Math.max((int) samples.getMax() * TIMEOUT_FACTOR, MIN_TIMEOUT_MILLIS), MAX_TIMEOUT_MILLIS);
+  }
+
+  @Override
+  public void close() {
+    if (closed.compareAndSet(false, true)) {
+      for (Callback callback : callbacks.values()) {
+        callback.completeExceptionally(new ConnectException());
+      }
+    }
+  }
+
+  /**
+   * Client connection callback.
+   */
+  final class Callback {
+    private final long id;
+    private final String type;
+    private final long time = System.currentTimeMillis();
+    private final long timeout;
+    private final ScheduledFuture<?> scheduledFuture;
+    private final CompletableFuture<byte[]> replyFuture;
+
+    Callback(long id, String type, Duration timeout, CompletableFuture<byte[]> future) {
+      this.id = id;
+      this.type = type;
+      this.timeout = getTimeoutMillis(type, timeout);
+      this.scheduledFuture = executorService.schedule(this::timeout, this.timeout, TimeUnit.MILLISECONDS);
+      this.replyFuture = future;
+      future.thenRun(() -> addReplyTime(type, System.currentTimeMillis() - time));
+      callbacks.put(id, this);
+    }
+
+    /**
+     * Returns the callback message type.
+     *
+     * @return the message type
+     */
+    String type() {
+      return type;
+    }
+
+    /**
+     * Fails the callback future with a timeout exception.
+     */
+    private void timeout() {
+      replyFuture.completeExceptionally(new TimeoutException("Request type " + type + " timed out in " + timeout + " milliseconds"));
```

Comment:
I think this method may need to remove the callback from `callbacks` here

Reply:
Yeah or risk of memory leak?